nextcloud/cms_pico
# Pico CMS for Nextcloud v1.0.0 (WIP)

:construction: **WORK IN PROGRESS** :construction: 

This is, without any question, a lot... :smile: This is a combination of a total refactoring and a rewrite from scratch - but IMHO **we're very close to Pico CMS for Nextcloud version 1.0.0** :confetti_ball: :tada:

**Testing the app** is pretty straight forward:

1. [Download the tarball/ZIP](https://github.com/PhrozenByte/nextcloud-pico/archive/pico-2.0.zip) or clone my fork (see [`PhrozenByte/nextcloud-pico`](https://github.com/PhrozenByte/nextcloud-pico/tree/pico-2.0), branch `pico-2.0`) to `apps/cms_pico/` of your Nextcloud instance (requires Nextcloud 14+)
2. Run `composer install`
3. Enable the app in your Nextcloud instance
4. Check out your personal and admin settings pages of "Pico CMS" 

**PRs to my fork and comments/reviews are very welcome**! :heart: My fork also allows edits from maintainers.

This PR solves all known (at least to me... I'll reference some issues in follow-up comments) issues with the current Nextcloud app, utilizes the latest release of Pico 2.1 (currently a dev build, however, I guarantee Pico 2.1 will be released before releasing this app) and I'm furthermore willing to officially support this app (Hi everyone, I'm Pico's lead dev :v: - i.e. Pico CMS for Nextcloud will be officially supported by Pico).

Please note that many commits are rather huge and don't really address single issues, but single aspects of the app. This is rather bad for change traceability, but IMHO contributes to the final solution, since it allowed me to ignore limitations of the existing code base. Just to mention that a thousand small commits with a tons of reverts won't be much better in regards of change traceability... So... :man_shrugging: However, naturally I didn't re-invent the wheel here and changed only a few concepts, not all of them :wink: Some commits do have some elaboration on my thoughts.

@daita you really did an amazing job here! Great work! :heart: I hope you forgive me for being so boldly doing this... You invested so much time, but priorities change (or, better to say I guess, the priorities of your employer change :see_no_evil:) and I felt like this app fell a bit behind :disappointed: So many people asked me about the plans for the Nextcloud app and I just could only answer "I don't know". Since it's an amazing project and heavily contributes to Pico's visibility, I didn't want to let it fall behind... And here we are. Thanks for all of your great work @daita! :+1:

---

**Screenshots:**

<details><summary>Example website</summary>

[![Example website](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/website.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/website.png)

</details>

<details><summary>List websites</summary>

[![List websites](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/list_websites.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/list_websites.png)

</details>

<details><summary>Create website</summary>

[![Create website](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/new_website.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/new_website.png)

</details>

<details><summary>Custom themes, plugins and templates</summary>

[![Custom themes, plugins and templates](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/custom.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/custom.png)

</details>

<details><summary>Webserver config</summary>

[![Webserver config](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/apache.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/apache.png)

</details>

---

**Some thoughts about some particular features:**

- Pico CMS for Nextcloud now supports both `mod_proxy`-based URLs (like `https://example.com/sites/my_pico_site`) and generic URLs (like `https://example.com/index.php/apps/cms_pico/pico/my_pico_site`). Currently the latter is shown by default (since it works out-of-the-box and webserver configuration was the most often asked question and source of problems until now) and the shorter form is just mentioned. I'll add a simple config option to enable short URLs after the admin has configured the webserver appropriately. - *Done* :+1:

- I'm planning to add a "Add custom plugin" feature just like for themes, allowing admins to add plugins to Pico. Plugins are tested for compatibility (Pico will throw an exception otherwise). We might even extend this to enabling plugins on a per-website basis. However, the latter is still a long way off. - *Done* :+1:

- Handling assets (content assets, theme assets and plugin assets) still isn't optimal (especially regarding browser caching). To solve the remaining issues I'll add a feature to Pico to support `%assets_url%` and `%plugins_url%` next to the already existing `%theme_url%` (+ corresponding Twig variables; actually this is already implemented in Pico 2.1). These variables can then be used by Pico CMS for Nextcloud to create version-specific URLs utilizing Nextcloud's ETag (content assets) resp. some random string that is generated when a theme/plugin is published or updated. Not sure about the details yet. - *Done* :+1:

- The two features listed above are currently just available for Pico plugins. I'll enable the same functionality for Pico themes, too. - *Done* :+1:

- I'm no big fan of unit tests (not because I don't like them, I usually just don't have time writing them...). All existing tests are still working, but I didn't write any new.

- I don't really know how Transifex works :confused: `de.js`, `de.json`, `de_DE.js`, `de_DE.json`, `en_GB.js` and `en_GB.json` are all fully up-to-date, but how can we import them to Transifex?

- I've absolutely no idea how this whole release process works (`.drone.yml` :question:)...
